New tutorial for the Gradle BT manual (#24913)
gradle/gradle: subprojects/docs/src/docs/userguide/tutorial/part4_gradle_plugins.adoc

// Copyright 2017 the original author or authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

[[part4_gradle_plugins]]
= Part 4: Apply Gradle Plugins and Distribute your App

> Learn the basics of Gradle's dependency management.

Author: Fixed in latest commit.

// Copyright 2017 the original author or authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

[[part4_gradle_plugins]]
= Part 4: Apply Gradle Plugins and Distribute your App

Learn about how Plugins extend Gradle's capabilities and use a popular plugin to distribute your app.

****
**In this section you will:**

- Apply a plugin
- Use the plugin
- Learn about distribution
- Explore other plugins
****

[[part4_begin]]
== Before you begin

1. Complete <<part3_gradle_dep_man#part3_begin,part 3>>.

== Step 1. Apply the Plugin

> A plugin is simply a set of tasks, available as packaged code that use the Gradle API to provide additional functionality and extend Gradle core.

Reviewer: ```suggestion
Plugins are the primary method of distributing custom tasks as packaged code.
They use the Gradle API to provide additional functionality and extend Gradle's core features.
```

> Applying a plugin to a project means that it allows the plugin to extend the project’s capabilities.

Reviewer: ```suggestion
Applying a plugin to a project executes code that can create tasks, configure properties, and otherwise extend the project's capabilities.
```

Plugins can:

> - Add tasks to the project (e.g. compile, test)

Reviewer: ```suggestion
- Add tasks to the project (e.g. compile, test).
```

- Extend the basic Gradle model (e.g. add new DSL elements that can be configured).

> - Configure the project, according to conversions (e.g. add new tasks or configure sensible defaults).

Reviewer: ```suggestion
- Configure the project according to conventions (e.g. add new tasks or configure sensible defaults).
```

- Apply specific configuration (e.g. add organisational repositories or enforce standards).
- Add new properties and methods to existing type via extensions.

> Our build file currently contains the `application` as the plugin:

Reviewer: ```suggestion
Our project currently applies a single plugin, the `application` plugin:
```

[source]
----
plugins {
    // Apply the application plugin to add support for building a CLI application in Java.
    application
}
----

There is a plugin that is maintained and distributed by Gradle called the <<distribution_plugin#distribution_plugin,Distribution Plugin>>.
The Distribution Plugin facilitates building archives that serve as distributions of the project.

Apply the plugin by adding `distribution` to the plugins block in `build.gradle.kts`:
[source]
----
plugins {
    // Apply the application plugin to add support for building a CLI application in Java.
    application

>     distribution

Reviewer: ❗ This is a bad example, as the application plugin automatically applies the disribution plugin: https://github.com/gradle/gradle/blob/f24d565784f6602c8b6c486c08fe5f4ee6dc76e9/subprojects/plugins/src/main/java/org/gradle/api/plugins/ApplicationPlugin.java#L75

Those distribution tasks will be available if you only apply application.

image::tutorial/intellij-idea-plugin.png[]

== Step 2. Use the Plugin
To use the plugin, simple run the `assembleDist` task by running `./gradlew :app:assembleDist`.
[source]
----
./gradlew :app:assembleDist

BUILD SUCCESSFUL in 577ms
----

The `assembleDist` task builds the project and creates a `app.tar` and `app.zip` distribution in the `build/distributions` directory.

image::tutorial/intellij-idea-dist.png[]

== Step 3. Distribute the App
A TAR or ZIP file can be used internally to distribute the app amount peers.

A distribution can also be published to the public using the <<publishing_ivy.adoc#publishing_ivy, Ivy Publish Plugin>> or the <<publishing_maven.adoc#publishing_maven, Maven Publish Plugin>>.

For example, if you wanted to publish your new app to an Ivy repository with the <<publishing_ivy.adoc#publishing_ivy, Ivy Publish Plugin>>, you would modify your build script as follows:
[source]
----
plugins {
    `ivy-publish`
}

publishing {
    publications {
        create<IvyPublication>("myDistribution") {
            artifact(tasks.distZip.get())
            artifact(tasks["customDistTar"])
        }
    }

> }

Reviewer: There is so much more to publishing, I think you should probably leave it out of this intro set of tutorials completely.
